File: java/src/main/java/com/refactoring/codesmells/couplers/Middleman.java
package com.refactoring.codesmells.couplers;

import java.util.*;

public class Middleman {

    public static void main(String[] args) {
        List<String> items = demoMiddleman();
        System.out.println("Items: " + items);
    }

    public static class Catalog {
        private Map<String, String> items = new HashMap<>();

        public void add(String id, String name) {
            items.put(id, name);
        }

        public Optional<String> find(String id) {
            return Optional.ofNullable(items.get(id));
        }

        public List<String> list() {
            return new ArrayList<>(items.values());
        }
    }

    public static class Shop {
        private Catalog catalog;

        public Shop(Catalog catalog) {
            this.catalog = catalog;
        }

        public void add(String id, String name) {
            catalog.add(id, name);
        }

        public Optional<String> find(String id) {
            return catalog.find(id);
        }

        public List<String> list() {
            return catalog.list();
        }
    }

    public static List<String> demoMiddleman() {
        Catalog c = new Catalog();
        Shop s = new Shop(c);
        s.add("1", "Book");
        s.add("2", "Pen");
        return s.list();
    }
}

File: python/src/code_smells/couplers/middleman.py
"""
Middleman Code Smell Example

A class exists solely to delegate most of its work to another class.
It adds no value and just forwards calls.

Refactoring suggestion: Remove Middle Man
"""


class Catalog:
    """The actual class doing the work"""

    def __init__(self) -> None:
        self._items: dict[str, str] = {}

    def add(self, item_id: str, name: str) -> None:
        self._items[item_id] = name

    def find(self, item_id: str) -> str | None:
        return self._items.get(item_id)

    def list(self) -> list[str]:
        return list(self._items.values())


class Shop:
    """
    Middleman: just delegates everything to Catalog.

    This class adds no value - it's just a pass-through.
    Clients should probably use Catalog directly.
    """

    def __init__(self, catalog: Catalog) -> None:
        self._catalog = catalog

    def add(self, item_id: str, name: str) -> None:
        self._catalog.add(item_id, name)

    def find(self, item_id: str) -> str | None:
        return self._catalog.find(item_id)

    def list(self) -> list[str]:
        return self._catalog.list()


def demo_middleman() -> list[str]:
    catalog = Catalog()
    shop = Shop(catalog)
    shop.add('1', 'Book')
    shop.add('2', 'Pen')
    return shop.list()
